video_core: Page manager/region manager optimization

LNDF opened the pull request:
Rewrite parts of the page manager and region manager to be faster

Use a bit array that uses AVX to accelerate.
Also tries to batch mprotects in the same region (another pr is needed to take advantage of this)

I have done zero testing right now with this and i rewrote parts of the memory tracker so there is a very high chance I broke something.

This is part of a series of PR with the hope of improving performance of games that use DMA, since it is currently quite slow (mainly because the ammount of memory protections and synced buffers).

LNDF committed:
Bit array test

LNDF committed:
Some corrections

LNDF committed:
Fix AVX path on SetRange

LNDF committed:
Finish bitArray

LNDF committed:
Batched protect progress

LNDF committed:
Inclusion fix

LNDF committed:
Last logic fixes for BitArray

LNDF committed:
Page manager: batch protect, masked ranges

LNDF committed:
Page manager bitarray

LNDF committed:
clang-format

LNDF committed:
Fix out of bounds read

raphaelthegreat commented on src/video_core/buffer_cache/region_manager.h:
```diff
@@ -0,0 +1,207 @@
+    VAddr cpu_addr = 0;
+    RegionBits cpu;
+    RegionBits gpu;
+    RegionBits untracked;
```
I recommend renaming untracked to write like in readbacks commit. It makes it a lot more clear what it does (set bits mean that page is writable, unset means write protected)

raphaelthegreat commented on src/video_core/buffer_cache/region_manager.h:
```diff
@@ -0,0 +1,207 @@
+        if constexpr (type == Type::GPU) {
+            mask &= ~untracked;
+        }
+        mask &= bits;
```
The if condition is actually not necessary and I had removed it in readbacks commit. It essentially keeps only the pages that are write protected, but if you mark a region as gpu modified all pages are write protected so shouldn't matter (assuming you handle gpu modified ranges properly, i.e unmarking them on write faults which isn't actually done on main, due to lack of readbacks)

Aside from that, you could possibly optimize this by fusing the SetRange and &= bits operations into a single one (could add a constructor to BitArray that receives another BitArray object and a range of pages to copy)

raphaelthegreat commented on src/video_core/buffer_cache/region_manager.h:
```diff
@@ -0,0 +1,207 @@
+        RegionBits mask = cpu ^ untracked;
+        untracked = cpu;
+        lk.unlock();
+        tracker->UpdatePageWatchersMasked<add_to_tracker>(cpu_addr, mask);
```
Does this make a difference between iterating the pages here and using normal update function? Assuming non contiguous mask, the gaps aren't gonna be affected so you will still do N = num contiguous regions, mprotect calls

LNDF replied:
Yes.
For example, in the following situation:
I want to untrack from page 0 to 15. But, since page 1 to 14 are already untracked, I end up only needing to untrack page 0 and 15.
With the previous implementation, UpdatePageWatchers would be called twoiceonce for page 0 and another time for page 15. 2 mprotects.
With the implementation on this PR if we want to untrack (unprotect) from page 0 to 15 and every page inbetween is not write protected, it will be handled as 1 single mprotect from page 0 to 15. You can see how that is handled in the UpdatePageWatchersMasked function.

raphaelthegreat replied:
I see that makes sense

LNDF committed:
clang

LNDF committed:
clang

LNDF committed:
Lock during callbacks

LNDF committed:
Rename untracked to writeable

LNDF committed:
Construct and mask in one step

LNDF committed:
Sync on region mutex for thw whole protection

This is a temporary workarround until a fix is found for the page manager having issues when multiple threads update the same page at the same time.

LNDF committed:
Bring back the gpu masking until properly handled

LNDF committed:
Merge branch 'main' into memory-tracker

LNDF committed:
Sync page manager protections

LNDF committed:
clang-format

LNDF committed:
Rename and fixups

LNDF committed:
I fucked up clang-formatting one more time...

LNDF committed:
kek
